File: test/org/stringtemplate/test/TestIndirectionAndEarlyEval.java
package org.stringtemplate.test;

import org.junit.Test;
import static org.junit.Assert.assertEquals;
import org.stringtemplate.ST;
import org.stringtemplate.STGroup;

import java.util.Map;
import java.util.HashMap;

public class TestIndirectionAndEarlyEval extends BaseTest {
    @Test public void testEarlyEval() throws Exception {
        String template = "<(name)>";
        ST st = new ST(template);
        st.add("name", "Ter");
        String expected = "Ter";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testIndirectTemplateInclude() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("foo", "bar");
        String template = "<(name)()>";
        group.defineTemplate("test", template);
        ST st = group.getInstanceOf("test");
        st.add("name", "foo");
        String expected = "bar";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testIndirectTemplateIncludeViaTemplate() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("foo", "bar");
        group.defineTemplate("tname", "foo");
        String template = "<(tname())()>";
        group.defineTemplate("test", template);
        ST st = group.getInstanceOf("test");
        String expected = "bar";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testIndirectProp() throws Exception {
        String template = "<u.(propname)>: <u.name>";
        ST st = new ST(template);
        st.add("u", new TestCoreBasics.User(1, "parrt"));
        st.add("propname", "id");
        String expected = "1: parrt";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testIndirectMap() throws Exception {
        STGroup group = new STGroup();
        group.defineTemplate("a", "[<it>]");
        group.defineTemplate("test", "hi <names:(templateName)()>!");
        ST st = group.getInstanceOf("test");
        st.add("names", "Ter");
        st.add("names", "Tom");
        st.add("names", "Sumana");
        st.add("templateName", "a");
        String expected =
            "hi [Ter][Tom][Sumana]!";
        String result = st.render();
        assertEquals(expected, result);
    }

    @Test public void testNonStringDictLookup() throws Exception {
        String template = "<m.(intkey)>";
        ST st = new ST(template);
        Map<Integer, String> m = new HashMap<Integer, String>();
        m.put(36, "foo");
        st.add("m", m);
        st.add("intkey", 36);
        String expected = "foo";
        String result = st.render();
        assertEquals(expected, result);
    }
}

File: objc/main/test/objc/TestIndirectionAndEarlyEval.m
#import "TestIndirectionAndEarlyEval.h"

@implementation TestIndirectionAndEarlyEval

- (void) testEarlyEval {
  NSString *template = @"<(name)>";
  ST *st = [[[ST alloc] init:template] autorelease];
  [st add:@"name" arg1:@"Ter"];
  NSString *expected = @"Ter";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testIndirectTemplateInclude {
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"foo" arg1:@"bar"];
  NSString *template = @"<(name)()>";
  [group defineTemplate:@"test" arg1:@"name" arg2:template];
  ST *st = [group getInstanceOf:@"test"];
  [st add:@"name" arg1:@"foo"];
  NSString *expected = @"bar";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testIndirectTemplateIncludeWithArgs {
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"foo" arg1:@"x,y" arg2:@"<x><y>"];
  NSString *template = @"<(name)({1},{2})>";
  [group defineTemplate:@"test" arg1:@"name" arg2:template];
  ST *st = [group getInstanceOf:@"test"];
  [st add:@"name" arg1:@"foo"];
  NSString *expected = @"12";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testIndirectTemplateIncludeViaTemplate {
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"foo" arg1:@"bar"];
  [group defineTemplate:@"tname" arg1:@"foo"];
  NSString *template = @"<(tname())()>";
  [group defineTemplate:@"test" arg1:@"name" arg2:template];
  ST *st = [group getInstanceOf:@"test"];
  NSString *expected = @"bar";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testIndirectProp {
  NSString *template = @"<u.(propname)>: <u.name>";
  ST *st = [[[ST alloc] init:template] autorelease];
  [st add:@"u" arg1:[[[User alloc] init:1 arg1:@"parrt"] autorelease]];
  [st add:@"propname" arg1:@"id"];
  NSString *expected = @"1: parrt";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testIndirectMap {
  STGroup *group = [[[STGroup alloc] init] autorelease];
  [group defineTemplate:@"a" arg1:@"x" arg2:@"[<x>]"];
  [group defineTemplate:@"test" arg1:@"names,templateName" arg2:@"hi <names:(templateName)()>!"];
  ST *st = [group getInstanceOf:@"test"];
  [st add:@"names" arg1:@"Ter"];
  [st add:@"names" arg1:@"Tom"];
  [st add:@"names" arg1:@"Sumana"];
  [st add:@"templateName" arg1:@"a"];
  NSString *expected = @"hi [Ter][Tom][Sumana]!";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

- (void) testNonStringDictLookup {
  NSString *template = @"<m.(intkey)>";
  ST *st = [[[ST alloc] init:template] autorelease];
  NSMutableDictionary *m = [[[NSMutableDictionary alloc] init] autorelease];
  [m setObject:36 arg1:@"foo"];
  [st add:@"m" arg1:m];
  [st add:@"intkey" arg1:36];
  NSString *expected = @"foo";
  NSString *result = [st render];
  [self assertEquals:expected arg1:result];
}

@end
